### PR TITLE
Adding a state 'unprepared' for establish claim tasks with no decision docs

### LINKED ISSUE #637: Caseworker:  EC Task:  Prefetch decision documents when creating Claims Establishment jobs
As a Caseworker, I only want an Establishment Claim (EC) task created when an appeal has an available decision document. Otherwise continue checking for the document each day.  So that no EC task is queued up without a decision document.

# Acceptance Criteria
1. Verify that an Establishment Claim Task is created only when an appeal has a decision document in VBMS.
2. Verify that when an appeal can not be associated with a decision document, the appeal is included in the next run of the task creation process.

# Technical Notes
When creating claims establishment jobs, we should prefetch the decision document from VBMS and store it in S3. 

Additionally, we should deal with three edge cases:
1) No decision document exists - in this case we should somehow mark this appeal as needing a decision so it can appear on the report described in #408
2) Two or more decision documents exist (in the time range) - A task should still be created, but these should marked as having too many decisions in the time range, and when that task is assigned to a user, they should be notified that they need to cancel the task and create the EP outside of caseflow.
3) Two or more appeals exist for the same veteran on the same decision date - do the same as edge case 2. Except mark each appeal as having a conflicting date with another appeal for that veteran. Inform the user that they must cancel the task and complete the EP Establishment outside of caseflow.

The extended solution for edge cases 2/3 will be fleshed out in future issues.

# Related Stories

### PR DESCRIPTION
**Pull request by sharonwarner**
This PR adds an 'unprepared' state for tasks. For Establish Claim tasks, we use that state to allow the system to search for decision docs. A job runs on all 'unprepared' tasks. When decision docs are found, the state switches to 'unassigned'. The added job will be cron-ed to run after CreateEstablishClaimTasksJob and ReassignOldTasksJob. 

Fixes #637 

- [ ] Added a unit test
- [ ] Unit tests pass
- [ ] UI works as expected

### CODE REVIEW

File: app/jobs/prepare_establish_claim_tasks_job.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+class PrepareEstablishClaimTasksJob < ActiveJob::Base
+  queue_as :default
+
+  def perform
+    EstablishClaim.where(aasm_state: "unprepared").each do |establish_claim|
+      establish_claim.prepare! if establish_claim.appeal.decision
```
Comment (mdbenjam) [inline app/jobs/prepare_establish_claim_tasks_job.rb]: I just made changes to this, that probably will break it. The method is now decisions, and I think returns a [] when there are no decisions.

File: app/jobs/prepare_establish_claim_tasks_job.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+class PrepareEstablishClaimTasksJob < ActiveJob::Base
```
Comment (joofsh) [inline app/jobs/prepare_establish_claim_tasks_job.rb]: Looks good 👍  Can you also add this task to the `config/sidekiq_cron.yml` so that it gets run every night? I think for this one we can actually remove the `disabled` part, won't hurt if it's already running in prod since we don't have any tasks yet.

File: app/jobs/prepare_establish_claim_tasks_job.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+  queue_as :default
+
+  def perform
+    EstablishClaim.where(aasm_state: "unprepared").each do |establish_claim|
```
Comment (joofsh) [inline app/jobs/prepare_establish_claim_tasks_job.rb]: Can we add a query method for this like the other states? `EstablishClaim.unprepared`

File: app/models/task.rb [hunk 4]
```diff
@@ -63,14 +63,12 @@ def completion_status_code(text)
   end
 
   aasm do
-    # state :unprepared, :unassigned, :assigned, :started, :completed
+    state :unprepared, initial: true
```
Comment (joofsh) [inline app/models/task.rb]: can we ad a comment explaining what `unprepared` means generally and in the specific case of EstablishClaim

File: spec/jobs/prepare_establish_claim_tasks_job_spec.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    }
+    @task_one = EstablishClaim.create(appeal: @appeal_one)
+    @task_two = EstablishClaim.create(appeal: @appeal_two)
+  end
+
+  context ".perform" do
+    it "prepares the correct tasks" do
+      expect(EstablishClaim.where(aasm_state: "unprepared").count).to eq(2)
+      PrepareEstablishClaimTasksJob.perform_now
+      expect(EstablishClaim.where(aasm_state: "unassigned").count).to eq(1)
```
Comment (mdbenjam) [inline spec/jobs/prepare_establish_claim_tasks_job_spec.rb]: Maybe ensure the correct task gets prepared, and the correct one is left unassigned. Rather than just looking at the counts.

File: app/jobs/prepare_establish_claim_tasks_job.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+      establish_claim.prepare! if establish_claim.appeal.decision
+    end
+  end
+end
```
Comment (joofsh) [inline app/jobs/prepare_establish_claim_tasks_job.rb]: I think the existing `CreateEstablishClaimTasksJob` will need to be updated with these changes too (can do in a separate PR if you prefer).

Now those tasks will get created in the `unprepared` state. We either need to check for an existing decision in that job too, or have this job run late enough so that it picks up tasks created earlier that night. Up to you which you think is easier/cleaner

File: spec/models/task_spec.rb [hunk 7]
```diff
@@ -74,17 +74,21 @@ class FakeTask < Task
 
     context ".assign!" do
       it "correctly assigns a task to a user" do
+        task.prepare!
```
Comment (mdbenjam) [inline spec/models/task_spec.rb]: Can you factor task.prepaer! into a before block?

File: app/jobs/prepare_establish_claim_tasks_job.rb (hunk 6 above)
Comment (mdbenjam) [inline app/jobs/prepare_establish_claim_tasks_job.rb]: I very much lean towards making this job run after that first one. It seems like CreateEstablishClaimTasksJob should create tasks even if there is no decision. I feel like we risk losing tasks otherwise.

Comment (sharonwarner) [inline app/jobs/prepare_establish_claim_tasks_job.rb]: Yeah, I lean towards that also. The reassign_old_tasks_job has the same problem. I think it's cleaner to have one (or two with the reassign job) job that creates all the tasks and then a second job that figures out the correct state of those tasks.

File: app/jobs/prepare_establish_claim_tasks_job.rb (hunk 1 above)
Comment (sharonwarner) [inline app/jobs/prepare_establish_claim_tasks_job.rb]: Yes! Merging master in broke it... I updated this to check the count of decisions.

File: app/jobs/prepare_establish_claim_tasks_job.rb (hunk 6 above)
Comment (joofsh) [inline app/jobs/prepare_establish_claim_tasks_job.rb]: 👍
